packages/salesforcedx-vscode-soql/src/client/client.ts

```diff
@@ -60,6 +59,28 @@ export async function startLanguageClient(
     synchronize: {
       configurationSection: 'soql',
       fileEvents: workspace.createFileSystemWatcher('**/*.soql')
+    },
+    middleware: {
```

Comment:
Can you provide an example on when this code is needed ? Which query and completions would I have to write to use this ?

Reply:
Not sure I understand the question... but let me try to give some context.

The connection to the Salesforce Org (to retrieve SObjects metadata, needed for completion) is already handled by the Extensions, where the LSP client runs.

So, we thought of some options:

1. Don't use LSP at all and implement completion inside the VSCode extension. Problem gone.
2. The LSP server could use its own separate connection. This would add some complexity to pass on the connection info/credentials from the extension to the server, making sure the right Org is hit (in sync with the extension) and prevent us from sharing caching.
3. The LSP server asks the client for metadata when needed. This would make each code-completion requests very chatty:
![image](https://user-images.githubusercontent.com/152839/103089752-b576b500-45cd-11eb-8bc2-27c32f4c907a.png)
4. The LSP server returns special completion items that the client must interpret and replace ("expand") with sobjects metadata:
![image](https://user-images.githubusercontent.com/152839/103089812-f1117f00-45cd-11eb-8acc-37fd14498d63.png)

So, we went with option 4, which allows us to continue using LSP while reusing the connections/caching of the extensions. What you see on this PR is the client-side of that.

More info here: https://salesforce.quip.com/JxYxAoa17YxO#NMUACAGfgxx

I hope my explanation makes sense. I would be happy to have a live discussion and go over more details. We already have improvements and refactorings coming on another PR on top of this one. Thanks!

Reply:
@dobladez could you give me access to that document ?

We should talk about how you are approaching the LSP functionality. Right now you are asking the client to massage/process the data returned by it. This might be ok for the SOQL extension but will make the SOQL LSP difficult to use in other scenarios/tools. At the moment, we know of at least two other tools outside of the SOQL extension where we want to use the language server and I don't believe the current approach is the correct one.

@praksb @jgellin-sf @dobladez lets talk about this once we're back from the holidays.

Reply:
Sure. Quip tells me you already have access to the quip doc....  anyway, the document doesn't say much more regarding connection-handling.

Reply:
circling back, we met and decided we'll change this in a future PR since it requires changes to the language server.